sycl/source/detail/program_manager/program_manager.hpp

```diff
@@ -383,13 +357,17 @@ class ProgramManager {
   /// in the sycl::detail::__sycl_service_kernel__ namespace which is
   /// exclusively used for this purpose.
   /// Access must be guarded by the m_KernelIDsMutex mutex.
-  std::unordered_set<std::string> m_ServiceKernels;
+  std::unordered_multimap<std::string, RTDeviceBinaryImage *> m_ServiceKernels;
 
   /// Caches all exported symbols to allow faster lookup when excluding these
   // from kernel bundles.
   /// Access must be guarded by the m_KernelIDsMutex mutex.
   std::unordered_set<std::string> m_ExportedSymbols;
 
+  /// Keeps all device images we are refering to during program lifetime. Used
+  /// for proper cleanup.
+  std::unordered_set<RTDeviceBinaryImageUPtr> m_DeviceImages;
```

Comment:
Non-blocking: why is this an `unordered_set` instead of a plain vector? Do we ever expect a duplicate image to be inserted here?